### PR TITLE
[DPriest] Typescript Conversion

### PR DESCRIPTION
**Pull request by TJMoats**
This one was a bit of a pain to convert. I fully expect there to be a ton of feedback on some of the typing XD

### CODE REVIEW

**Comment by TJMoats** [comment]
@karlpralow I've converted the module that you just made over to tsx. You might want to verify that I didn't bug anything up.

File: src/parser/priest/discipline/modules/core/Channeling.ts
Comment (rp4rk) [inline src/parser/priest/discipline/modules/core/Channeling.ts:14]: Can these (and similar occurrences) not be handled by type inference?

File: src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx
Comment (rp4rk) [inline src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx]: Was this `Math.round` causing problems?

File: src/parser/priest/discipline/modules/spells/SinsOfTheMany.tsx
Comment (rp4rk) [inline src/parser/priest/discipline/modules/spells/SinsOfTheMany.tsx:25]: The object here is more legible for looking at target count to bonus damage mapping

File: tsconfig.json
Comment (rp4rk) [inline tsconfig.json]: What are the implications of this configuration change?

**Review by rp4rk — APPROVED** [review]
Looks good overall, I have a few questions but I don't think they're blockers

File: src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx
Comment (Dambroda) [reply src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx]: formatPercentage would return a string, so Math.round which takes a number would have been angry. The way to fix it depends on what the end goal is.. but I'm pretty sure in this case the correct answer is to provide the second argument to the format function to trim off the decimal places

File: src/parser/core/EventGrouper.ts
Comment (Dambroda) [inline src/parser/core/EventGrouper.ts]: `cache: { [stem: number]: Event<any>[] }`

File: src/parser/core/Module.ts
Comment (Dambroda) [inline src/parser/core/Module.ts]: What required this change?

File: src/parser/priest/discipline/modules/Abilities.ts
Comment (Dambroda) [inline src/parser/priest/discipline/modules/Abilities.ts]: At the top of the file,
`spellbook(): SpellbookAbility[] {` 

Will both typecheck the abilities in the file and also automatically infer the `haste` type

File: src/parser/shared/modules/CooldownThroughputTracker.tsx
Comment (Dambroda) [inline src/parser/shared/modules/CooldownThroughputTracker.tsx]: This is blocking, please remove this and we can work to fix whatever issue it was added to resolve

File: src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts
Comment (Dambroda) [inline src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts]: This looks at all events, its `Event<any>` (same with fixedEvents declaration)

File: src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts
Comment (Dambroda) [inline src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts]: Instead, because it's actually added in a lot of normalizers, you should add `__modified?: boolean` to the Event interface right below `__fabricated?: boolean`

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (Dambroda) [inline src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: __modified should be added to the base Event interface (see previous comment)

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (Dambroda) [inline src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: `Event<any>`

**Review by Dambroda — CHANGES_REQUESTED** [review]
I got through about half of the files so far, but I figured I'd submit the first part so you can get a head start. Also, I can't leave a comment on it because it there were no line changes, but also please add the typing to CONFIG like Holy  :)

File: src/parser/priest/discipline/modules/spells/Penance.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Penance.tsx]: (referring to this comment in other comments)

```typescript
export function IsPenanceDamageEvent(event: DamageEvent): event is PenanceDamageEvent {
  return (event as PenanceDamageEvent).penanceBoltNumber !== undefined;
}
```
 and the same for the heal event

https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

File: src/parser/priest/discipline/modules/spells/Penance.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Penance.tsx]: `DamageEvent`

File: src/parser/priest/discipline/modules/spells/Penance.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Penance.tsx]: `(event as PenanceDamageEvent).penanceBoltNumber`

(same last two comments for the heal listener)

File: src/parser/priest/discipline/modules/spells/Castigation.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Castigation.tsx]: Just a `DamageEvent` (cant assume all events grabbed by this listener are penance events)

The way you can typecheck things like this is with typeguards (refer to previous comment)
```typescript
on_byPlayer_damage(event: DamageEvent) {
  if(!IsPenanceDamageEvent(event)) {
    return;
  }
  // Everything after that check will automatically cast 'event' to PenanceDamageEvent

  // If you trust your other modules to only add the penanceBoltNumber property to penance events you could technically even remove the event.ability.guid check here now
  if (event.ability.guid !== SPELLS.PENANCE.id || event.penanceBoltNumber !== 3) {
      this._isCastigationBolt = false;
      return;
    }
```
Same with the Heal event listener

File: src/parser/priest/discipline/modules/spells/Evangelism.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Evangelism.tsx]: `return Object.keys(this._evangelismStatistics).map(Number).map((key: number) => this._evangelismStatistics[key]);`

It's annoying, but Object.keys always returns a string array regardless of the TypeScript index type. You can map the keys to number pretty safetly by just adding a `.map(Number)`

File: src/parser/core/Module.ts
Comment (TJMoats) [reply src/parser/core/Module.ts]: There are non-analyzer modules that have the analyzer passed down to them. The Atonement Healing Breakdown uses this variable, but it pulls it from the parent that creates the module. There are a number of modules that do something similar, and I didn't see too much of a stink from making it public instead of protected. I can modify that module to make it fetch the owner some other way if you think I should.

File: src/parser/priest/discipline/modules/core/Channeling.ts
Comment (TJMoats) [reply src/parser/priest/discipline/modules/core/Channeling.ts:14]: I just find that it's cleaner to add the types. They can be inferred if you think I should remove the types. Mind you the disc modules are a bit of a mixed bag for typing modules. There are a ton of nulls that end up being booleans, so I added these in a lot of places to maintain my sanity :D

File: src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx
Comment (TJMoats) [reply src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx]: What he said ^^ formatPercentage was returning a string.

File: src/parser/priest/discipline/modules/spells/SinsOfTheMany.tsx
Comment (TJMoats) [reply src/parser/priest/discipline/modules/spells/SinsOfTheMany.tsx:25]: objects being indexed with ints is usually a pretty gangly practice imo. I'd almost rather make it into a function instead of a purely int indexed object. Mind you there are objects in these modules that are indexed by objects so /shrug

I can change it if you really think it should remain an object.

File: tsconfig.json
Comment (TJMoats) [reply tsconfig.json]: One of the modules uses a custom class that someone wrote that allows you to call it's members like an array. It's the EventGrouper class. I didn't 100% get why it needed to be like that, but this change is necessary for typescript to allow iteration in that class.

Comment (TJMoats) [reply tsconfig.json]: I'd also 100% be down to axe that class, disc is the only spec that uses it as far as I can tell.

File: src/parser/shared/modules/CooldownThroughputTracker.tsx
Comment (TJMoats) [reply src/parser/shared/modules/CooldownThroughputTracker.tsx]: Yeah, I didn't really know how to handle this one. I have no idea what the module I changed this for was trying to accomplish.

https://github.com/WoWAnalyzer/WoWAnalyzer/blob/93f86fead27addf6aec9a2d5860f6e43f246bb45/src/parser/priest/discipline/modules/features/CooldownThroughputTracker.ts#L43

I was trying to fix the typing here ^^ 

I'd LOVE some additional insight into what this is actually doing.

File: src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts
Comment (TJMoats) [reply src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts]: The base event doesn't have ability: Ability which is what is being used in this module. That's why I manually added the possible types.

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (TJMoats) [reply src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: Again, this module uses things that exist in ModifiedDamageEvent | ModifiedHealEvent, but not in the base Event<any>. How should I go about typing that?

File: src/parser/priest/discipline/modules/spells/Penance.tsx
Comment (TJMoats) [reply src/parser/priest/discipline/modules/spells/Penance.tsx]: I can't typecheck this at function call time as the field is added 100% in this module. Are you saying that I should add IsPenanceDamageEvent typechecking for the other modules that use the penanceBoltNumber?

File: src/parser/priest/discipline/modules/spells/Penance.tsx
Comment (TJMoats) [reply src/parser/priest/discipline/modules/spells/Penance.tsx]: I can't use the typechecking to guard this function, this function is what adds the field to the event.

File: src/parser/core/EventGrouper.ts
Comment (TJMoats) [reply src/parser/core/EventGrouper.ts]: Fixed!

File: src/parser/priest/discipline/modules/Abilities.ts
Comment (TJMoats) [reply src/parser/priest/discipline/modules/Abilities.ts]: Fixed!

File: src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts
Comment (TJMoats) [reply src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts]: Fixed!

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (TJMoats) [reply src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: Fixed!

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (TJMoats) [reply src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: I've added casting to every place that you made this request. Lemme know if I should go down a different route!

File: src/parser/priest/discipline/modules/spells/Penance.tsx
Comment (TJMoats) [reply src/parser/priest/discipline/modules/spells/Penance.tsx]: Ignore that, I get it now :D

File: src/parser/priest/discipline/modules/spells/Evangelism.tsx
Comment (TJMoats) [reply src/parser/priest/discipline/modules/spells/Evangelism.tsx]: Fixed!

File: src/parser/shared/modules/CooldownThroughputTracker.tsx
Comment (TJMoats) [reply src/parser/shared/modules/CooldownThroughputTracker.tsx]: For now I've commented out the code in that module

File: tsconfig.json
Comment (rp4rk) [reply tsconfig.json]: I think I authored it back when it was looking like the only sane way of handling penance and Atonement was to group events within a threshold, but from the looks of things `downLevelIteration` brings typescript closer inline as a superset of ES so I'm not certain here. I'll keep this in mind as I'm working through the Discipline stuff for shadowlands and see how it pans out.

File: src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts
Comment (Dambroda) [reply src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts]: I have already added a type guard for this exact sitch! 
```typescript
     if(!HasAbility(event)) {
        return;
      }
```
Can add that right after the first `fixedEvents.push(event)` and all of the other instances of event.ability shouldn't error. Your casting is also fine in this case.

The reason this is important is because by telling the method that the arguments are of a certain type, the scope of the function will take your word as law. The way you had it originally, you could technically write `console.log(event.ability.guid)` on L20, before any of the `event.type` checks, and typescript would say that's fine. However, because `normalize` actually looks at all events, not only ability events, you would always get a crash that says `event.ability` is undefined when you try to access `guid`.

Technically resolved conversation, I'll leave casting vs typeguard up to you

File: src/parser/shared/modules/CooldownThroughputTracker.tsx
Comment (Dambroda) [reply src/parser/shared/modules/CooldownThroughputTracker.tsx]: I had to look into this,

image

it looks like its perfectly valid, and used in interface/others/Cooldown.js L 294-297

In the core CooldownThroughputTracker (i.e. the file this comment was originally for):
```typescript
export type SummaryDef = {
  label: string,
  tooltip: string,
  value: number | string,
}
export type CooldownSpell = {
  spell: any,
  summary: Array<BUILT_IN_SUMMARY_TYPES | SummaryDef>,
  startBufferFilter?: EventFilter<any>,
  startBufferMS?: number,
  startBufferEvents?: number,
  petID?: number,
};
```

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (Dambroda) [inline src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: Just as a comment, I am going to PR a fix for this so that you don't have to cast this every time. I'll @ you to see the fix

File: src/parser/priest/discipline/modules/spells/Schism.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Schism.tsx]: This probably needs to be cleaned up; right now the event itself is being used to index the object, could that be changed to index on the events timestamp?

File: src/parser/priest/discipline/modules/spells/Atonement.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Atonement.tsx:106]: I believe these (the timestamps) should stay, and the TS ignore can be moved up. I didn't realize how badly the shape was for the EventEmitter, but once that's fixed the ignores shouldn't be needed

File: src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx
Comment (Dambroda) [reply src/parser/priest/discipline/modules/features/PurgeTheWicked.tsx]: Do you know if you want to keep the decimals? If you do `formatPercentage(recommended, 0)` it will look the same as the previous Math.round (I believe)

File: src/parser/priest/discipline/modules/features/Checklist/Component.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/features/Checklist/Component.tsx]: `combatant: Combatant, castEfficiency: CastEfficiency, thresholds: any` (im pretty sure I have the thresholds type somewhere added but I think it requires going through *every* one of your thresholds and making sure they are correct.. wont block this PR for that)

File: tsconfig.json
Comment (Dambroda) [reply tsconfig.json]: Was this resolved? I wasn't sure if the final decision was to actually change tsconfig or to add some tsignores where the different iterables were used

**Comment by Dambroda** [comment]
Conflicts, that one question about tsconfig (wary about changing anything there), and then LGTM :)

File: tsconfig.json
Comment (TJMoats) [reply tsconfig.json]: I think the options are

1) Change the config
2) Rewrite the module

I don't think that this is something that will ts-ignore well.

Comment (Dambroda) [reply tsconfig.json]: @rp4rk From what you looked into is this config safe to merge in?

**Comment by TJMoats** [comment]
@Dambroda 

I reverted the modules that were being pains back to js. 
I reverted the config change. 
I reverted the public/protected change.

File: src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts
Comment (Dambroda) [inline src/parser/priest/discipline/normalizers/PowerWordRadianceNormalizer.ts]: With new AnyEvent you can remove these casts now :D

File: src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts
Comment (Dambroda) [reply src/parser/priest/discipline/normalizers/AtonementSuccessiveDamage.ts]: I've put this fix in, AnyEvent now doesn't require the casts in this file (`as XEvent` should all be able to remove)

File: src/parser/priest/discipline/modules/spells/Penance.js
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Penance.js]: unneeded parenthesis

File: src/parser/priest/discipline/modules/spells/Contrition.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/spells/Contrition.tsx]: semicolon

File: src/parser/priest/discipline/modules/features/AtonementHealingDone.tsx
Comment (Dambroda) [inline src/parser/priest/discipline/modules/features/AtonementHealingDone.tsx]: accidental extra space

**Review by Dambroda — APPROVED** [review]
Few small formatting things that got messed up with the merge, but I am approving this overall. Fix any/none, merge, and then in the next few weeks I'll take the second pass through disc and finish the conversions and resolve any of the leftover comments

**Comment by TJMoats** [comment]
Changes made for that feedback!

**Pull request merged**
